core/shared/src/main/scala/zio/Queue.scala

```diff
@@ -286,79 +287,91 @@ object Queue {
+      takers: ConcurrentLinkedDeque[Promise[Nothing, A]],
       isShutdown: AtomicBoolean
     )(implicit trace: Trace): UIO[Boolean]
 
     def unsafeOnQueueEmptySpace(
       queue: MutableConcurrentQueue[A],
-      takers: MutableConcurrentQueue[Promise[Nothing, A]]
+      takers: ConcurrentLinkedDeque[Promise[Nothing, A]]
     ): Unit
 
     def surplusSize: Int
 
     def shutdown(implicit trace: Trace): UIO[Unit]
 
+    @tailrec
     final def unsafeCompleteTakers(
       queue: MutableConcurrentQueue[A],
-      takers: MutableConcurrentQueue[Promise[Nothing, A]]
-    ): Unit = {
-      // check if there is both a taker and an item in the queue, starting by the taker
-      var keepPolling = true
-      val nullTaker   = null.asInstanceOf[Promise[Nothing, A]]
-      val empty       = null.asInstanceOf[A]
-
-      while (keepPolling && !queue.isEmpty()) {
-        val taker = takers.poll(nullTaker)
-        if (taker eq nullTaker) keepPolling = false
-        else {
-          queue.poll(empty) match {
-            case null =>
-              unsafeOfferAll(takers, taker +: unsafePollAll(takers))
-            case a =>
-              unsafeCompletePromise(taker, a)
-              unsafeOnQueueEmptySpace(queue, takers)
+      takers: ConcurrentLinkedDeque[Promise[Nothing, A]]
+    ): Unit =
+      if (!takers.isEmpty && draining.compareAndSet(false, true)) {
+        try {
+          var keepPolling      = true
+          val empty            = null.asInstanceOf[A]
+          var notifyEmptySpace = false
+          while (keepPolling) {
+            val taker = takers.poll()
+            if (taker eq null) keepPolling = false
+            else {
+              queue.poll(empty) match {
+                case null =>
+                  takers.addFirst(taker)
+                  keepPolling = false
+                case a =>
+                  unsafeCompletePromise(taker, a)
+                  notifyEmptySpace = true
+              }
+            }
           }
-          keepPolling = true
+          if (notifyEmptySpace) unsafeOnQueueEmptySpace(queue, takers)
+        } finally {
+          draining.set(false)
         }
+
+        // We need to check in case someone added a putter or pulled from the queue since our last check
+        // while we were still holding the lock
+        if (!queue.isEmpty()) unsafeCompleteTakers(queue, takers)
       }
-    }
+
   }
 
   private object Strategy {
 
     final case class BackPressure[A]() extends Strategy[A] {
+      private[this] val notifying = new AtomicBoolean(false)
+
       // A is an item to add
       // Promise[Nothing, Boolean] is the promise completing the whole offerAll
       // Boolean indicates if it's the last item to offer (promise should be completed once this item is added)
-      private val putters = MutableConcurrentQueue.unbounded[(A, Promise[Nothing, Boolean], Boolean)]
+      private val putters = new ConcurrentLinkedDeque[(A, Promise[Nothing, Unit], Boolean)]
```

Comment:
Note that we use a `ConcurrentLinkedDeque` so that we can efficiently place an item back at the start of the queue without pulling the entirety of it only to add it back again